backend/src/middleware/handleNotifications/handleNotificationsMiddleware.js

```diff
@@ -0,0 +1,156 @@
+import { UserInputError } from 'apollo-server'
+import extractMentionedUsers from './notifications/extractMentionedUsers'
+import extractHashtags from './hashtags/extractHashtags'
+
+const notifyUsers = async (label, id, idsOfUsers, reason, context) => {
+  if (!idsOfUsers.length) return
+
+  // Done here, because Neode validation is not working.
+  const reasonsAllowed = ['mentioned_in_post', 'mentioned_in_comment', 'comment_on_post']
+  if (!reasonsAllowed.includes(reason)) {
+    throw new UserInputError('Notification reason is not allowed!')
+  }
+  if (
+    (label === 'Post' && reason !== 'mentioned_in_post') ||
+    (label === 'Comment' && !['mentioned_in_comment', 'comment_on_post'].includes(reason))
+  ) {
+    throw new UserInputError('Notification fits not to reason!')
+  }
+
+  const session = context.driver.session()
+  const createdAt = new Date().toISOString()
+  let cypher
+  switch (reason) {
+    case 'mentioned_in_post': {
+      cypher = `
+        MATCH (post: Post { id: $id })<-[:WROTE]-(author: User)
+        MATCH (user: User)
+        WHERE user.id in $idsOfUsers
+        AND NOT (user)<-[:BLOCKED]-(author)
+        CREATE (notification: Notification {id: apoc.create.uuid(), read: false, reason: $reason, createdAt: $createdAt })
+        MERGE (post)-[:NOTIFIED]->(notification)-[:NOTIFIED]->(user)
+      `
+      break
+    }
+    case 'mentioned_in_comment': {
+      cypher = `
+        MATCH (postAuthor: User)-[:WROTE]->(post: Post)<-[:COMMENTS]-(comment: Comment { id: $id })<-[:WROTE]-(author: User)
+        MATCH (user: User)
+        WHERE user.id in $idsOfUsers
+        AND NOT (user)<-[:BLOCKED]-(author)
+        AND NOT (user)<-[:BLOCKED]-(postAuthor)
+        CREATE (notification: Notification {id: apoc.create.uuid(), read: false, reason: $reason, createdAt: $createdAt })
+        MERGE (comment)-[:NOTIFIED]->(notification)-[:NOTIFIED]->(user)
+      `
+      break
+    }
+    case 'comment_on_post': {
+      cypher = `
+        MATCH (postAuthor: User)-[:WROTE]->(post: Post)<-[:COMMENTS]-(comment: Comment { id: $id })<-[:WROTE]-(author: User)
+        MATCH (user: User)
+        WHERE user.id in $idsOfUsers
+        AND NOT (user)<-[:BLOCKED]-(author)
+        AND NOT (author)<-[:BLOCKED]-(user)
+        CREATE (notification: Notification {id: apoc.create.uuid(), read: false, reason: $reason, createdAt: $createdAt })
+        MERGE (comment)-[:NOTIFIED]->(notification)-[:NOTIFIED]->(user)
+      `
```

Comment:
Are the `mentioned_in_comment` and `comment_on_post` cypher statements the same, or am I missing something?
the `$reason` would be different, right? but since it's passed in as a variable that would be ok to use the same cypher statement, no?

Reply:
oh, this is different, no
```
AND NOT (user)<-[:BLOCKED]-(author)
AND NOT (author)<-[:BLOCKED]-(user)
```
right?

Reply:
Yes, this is different. 😉